### DISCUSSION TITLE
Online playground

### DISCUSSION THREAD

**Discussion by hatemhosny** [AmrDeveloper/GQL#98] [Ideas]
السلام عليكم و رحمة الله و بركاته 

This is such an impressive project. Great work.
I have been following your work for a while and have seen your [interview with Ahmed Elemam](https://youtu.be/ZP4_F_Nz0XQ?si=7LUSLEcaz0kjrTap).

I'm the author of [LiveCodes](https://livecodes.io), a [feature-rich](https://livecodes.io/docs/features/), [open-source](https://github.com/live-codes/livecodes), [client-side](https://livecodes.io/docs/why/#client-side), code playground that supports [80+ languages and frameworks](https://livecodes.io/docs/why/#language-support). You can find the [docs here](https://livecodes.io/docs/). The [announcement post](https://blog.livecodes.io/introducing-livecodes/) gives a general overview. [This video](https://youtu.be/QHGSOcOUeuU?si=mk2nTO2vpPI_nQNT) (also with Ahmed Elemam 😊) gives a general idea.

I was thinking about adding support for GQL in LiveCodes. This would have some benefits including:
- A low barrier to entry. People can easily try it online without having to download or install anything.
- A nice editing experience in Monaco editor (the code editor that powers VS code). I think SQL syntax highlighting will be very suitable. This is a demo running SQLite: https://livecodes.io/?template=sql, and this runs PostgreSQL: https://livecodes.io/?template=postgresql. Check all starter templates here: https://livecodes.io/?new (please note that the app is client-side only, with no backend. Everything runs in the browser)
- Code in LiveCodes can be [shared](https://livecodes.io/docs/features/share/), [exported](https://livecodes.io/docs/features/export/) (e.g. GitHub gist) and [deployed](https://livecodes.io/docs/features/deploy/) (to GitHub Pages).
- Playgrounds can be [embedded](https://livecodes.io/docs/features/embeds) in any webpage, using a powerful, yet easy-to-use, [SDK](https://livecodes.io/docs/sdk).
- There is no need for servers to provide this. No running costs.

However, there are some challenges:
- Being a client-side only app, we have to run the compiler in the browser. I believe GQL is written in Rust. Can we provide a wasm version of it that can run in the browser. (P.S. I have no experience with Rust)
- I see GQL queries local file system. This is not available in the browser.So can we have something like a virtual file system on which the files would be present.
- How would we load the files? Would it be loaded from local file system of the user? Or even better a remote git repo? Either clone the repo to the virtual file system or even run against GitHub API.

I would like to know what you think.
Any way, such a great project. Congratulations 💚

**Comment by AmrDeveloper** [comment]
عليكم السلام ورحمه الله وبركاته

Thank you Dr Hatem, I know your amazing project from interview with Ahmed Elemam too,

I am interested to add GQL to LiveCodes, i think i can provide a WASM version with every release but we need to do some test and experment the integration.

There are some limitation if we want to work with Github API, for example you will face a API limits before loading all commits, so i am thinking of two options, first we can search if we can run on local repository from FileSystem or we can mock a data and load it to the engine so the user can run any query on this data, this idea can work with Git, FileSystem, AST...etc

**Comment by hatemhosny** [comment]
That's great.

I agree with the API limitations. 

This project allows running git in the browser with a virtual file system, including cloning a whole repo.
https://github.com/isomorphic-git/isomorphic-git

May be we consider such approach.

**Comment by hatemhosny** [comment]
The File System Access API may also be useful, inspite of relatively limited browser support.
https://developer.chrome.com/docs/capabilities/web-apis/file-system-access
https://developer.mozilla.org/en-US/docs/Web/API/FileSystem

Another web API that might be useful is: The origin private file system
https://web.dev/articles/origin-private-file-system
This was recently used by [PhotoShop to make a web version](https://medium.com/@addyosmani/photoshop-is-now-on-the-web-38d70954365a).

**Reply by AmrDeveloper** [reply]
This is a good option too, i will try to build a wasm executable and learn how to config it to use System files API
